clients/src/main/java/org/apache/kafka/common/utils/ByteBufferOutputStream.java

```diff
@@ -105,7 +105,7 @@ public int initialCapacity() {
         return initialCapacity;
     }
 
-    private void maybeExpandBuffer(int remainingRequired) {
+    public void maybeExpandBuffer(int remainingRequired) {
```

Comment:
Since what you are proposing is a larger refactor, I suggest we leave it for future work. Since `ByteBufferOutputStream` is advertised as an expandable stream on top of `ByteBuffer`, I personally do not see the harm in allowing the caller to expand it manually. In addition to allowing for use cases which require direct access to a `ByteBuffer`, it potentially enables more efficient re-allocation where the caller knows exactly how many bytes will be written. For example, if I rewrote `writeHeader` to use `ByteBufferOutputStream`, the fields would be written individually which may call for multiple re-allocations. To get around that, I would probably use the `position` hack above to force expansion, but it's easier to have an API to do it directly.